SAMtags.tex

```diff
@@ -453,6 +453,165 @@ \subsubsection{Color space}
+
+% MP was the former quality score for MM.  However being Phred scores
+% it can only reasonable record probabilities for highly likely
+% events, making it inappropriate for callers (eg ONT's) that wish to
+% jointly call probabilities for the entire trained set of
+% possibilities.  We could use log-odds, similar to how early Illumina
+% runs did to record likelihoods for A, C, G and T irrespective of
+% call, but for now we're using linear-scaled probabilities.  These
+% are in the ML tag.
+%
+% The MP tag is left here for now as the jury is still out on whether
+% we'll need it in the future.
+%
+% \item[MP:Z:\tagvalue{qualities}]
+% \hfill\\
+% The optional {\tt MP} tag lists the Phred qualities of each modification listed in the {\tt MM} tag in the order they occur.
+% The qualities are encoded in the same manner as the primary {\sf QUAL} field; one byte per quality with ASCII value Phred score + 33.
+% A space character (`{\tt \textvisiblespace}') should be used as a separator between concatenated quality strings when multiple modification lists are present in the {\tt MM} tag.
+% The length should match the number of position deltas from {\tt MM} plus 1 per space character required.
+%
+% For example ``{\tt MM:Z:C+m,5,12,3;C+h,57;}'' may have an associated
+% quality tag of ``{\tt MP:Z:5EB /}''.
+%
+% Where multiple modification types are listed together, such as in ``{\tt MM:Z:C+mh,5,12,3;}'' the quality values are interleaved in order ({\tt m} at 6, {\tt h} at 6, {\tt m} at 19, {\tt h} at 19 and so on), giving 6 quality values in total for this example.
+%
+% Quality values for ambiguity codes give the likelihood that the
+% modification is one of the possible codes compatible with that
+% ambiguity code.  For example {\tt MM:Z:C+C,10; MP:Z:+} indicates a C
+% call with an unspecified modification and the phred score of 10 (ASCII
+% value {\tt +}).  This corresponds to a 90\% chance of the base being
+% modified.
+%
+% To represent several possible modifications at the same site the {\tt MP} tag can be used to indicate the probabilities of each possibility.
+% The values used should be absolute probabilities, not relative between the alternatives.
+% For example, a C base that has 95\% chance of being modified with 5mC being three times more likely than 5hmC will encode 5mC with 67.5\% probability ($0.9 * 0.75$ giving phred score 5, ASCII value {\tt \&})and 5hmC with 22.5\% probability ($0.9 * 0.25$ giving phred score 1, ASCII value {\tt "}).
+% This could be represented with ``{\tt MM:Z:C+m,10;C+h,10; MP:Z:" \&}''.
+
+\item[Ml:B:C,\tagvalue{scaled-probabilities}]
+\hfill\\
+The optional {\tt Ml} tag lists the probability of each modification listed in the {\tt Mm} tag being correct, in the order that they occur.
+The continuous probability range 0.0 to 1.0 is remapped in equal
+sized portions to the discrete integers 0 to 255 inclusively. Thus the
+probability range corresponding to integer value $N$ is $N/256$ to
+$(N+1)/256$.
```

Comment:
so, we can say that a modification happens at 100% but we cannot state that a modification definitely did _not_ occur? it might be useful to be able to express a certainty that a modification did not occur, _n'est pas_?

Reply:
I suggest we make this $p=N/255$ instead.

Reply:
This definition defines N -> p. What is the corresponding formula for of p -> N?

Reply:
I brought this up in a [thread earlier](https://github.com/samtools/hts-specs/pull/418#pullrequestreview-467822286). It seems the most logical thing to do would be to bin the space from 0 to 1 in equal sized bins. Thus we can't represent exactly 0.0 or 1.0, but instead represent the range 0-1/256 and 255/256-1. 

It seems the only other logical conversion would be to have special cases for 0.0 and 1.o probabilities, and then have the rest of the range equally spaced. But as noted by @jkbonfield the need for absolute certainty seems limited.

Reply:
> 
> 
> so, we can say that a modification happens at 100% but we cannot state that a modification definitely did _not_ occur? it might be useful to be able to express a certainty that a modification did not occur, _n'est pas_?

Actually we can specify that.  We list the modification code and give it $p=0$ to state the likelihood of this C being m is zero.  Infact this is basically what ONT already do, albeit possibly with a low cutoff to avoid excessive bloat.

Are you suggesting instead we take the opposite approach of eg `C+C,0,0,3,0,1;` to state that these Cs are real C and not modified ones?  I'm not sure I see the merit in that.

Reply:
I think I was confused about how the formula for N(p) is defined. I think we should write it explicitly: 
```
N=floor(p*255) 
```

Reply:
I agree adding the formula would help with clarity. It might be good to catch for the edge case where `p=1.0`. E.g. `N=min(255, floor(p*256))`